src/threads_mngr.c

```diff
@@ -0,0 +1,197 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "threads_mngr.h"
+/* Anti-warning macro... */
+#define UNUSED(V) ((void) V)
+
+#ifdef __linux__
+#include "zmalloc.h"
+#include "atomicvar.h"
+#include "server.h"
+
+#include <signal.h>
+#include <time.h>
+#include <errno.h>
+#include <semaphore.h>
+#include <sys/syscall.h>
+
+#define IN_PROGRESS 1
+static const clock_t RUN_ON_THREADS_TIMEOUT = 2;
+
+/*================================= Globals ================================= */
+
+static run_on_thread_cb g_callback = NULL;
+static volatile size_t g_tids_len = 0;
+static void **g_output_array = NULL;
+static redisAtomic size_t g_thread_ids = 0;
+static redisAtomic size_t g_num_threads_done = 0;
+
+static sem_t wait_for_threads_sem;
+
+/* This flag is set while ThreadsManager_runOnThreads is running */
+static redisAtomic int g_in_progress = 0;
+
+/*============================ Internal prototypes ========================== */
+
+static void invoke_callback(int sig);
+/* returns 0 if it is safe to start, IN_PROGRESS otherwise. */
+static int test_and_start(void);
+static void wait_threads(void);
+/* Clean up global variable. 
+Assuming we are under the g_in_progress protection, this is not a thread-safe function */
+static void ThreadsManager_cleanups(void);
+
+/*============================ API functions implementations ========================== */
+
+void ThreadsManager_init(void) {
+    /* Register signal handler */
+    struct sigaction act;
+    sigemptyset(&act.sa_mask);
+    /* Not setting SA_RESTART flag means that If a signal handler is invoked while a
+    system call or library function call is blocked, use the default behavior
+    i.e., the call fails with the error EINTR */
+    act.sa_flags = 0;
+    act.sa_handler = invoke_callback;
+    sigaction(SIGUSR2, &act, NULL);
+}
+
+void **ThreadsManager_runOnThreads(pid_t *tids, size_t tids_len, run_on_thread_cb callback) {
+    /* Check if it is safe to start running. If not - return */
+    if(test_and_start() == IN_PROGRESS) {
+        return NULL;
+    }
+
+    /* Update g_callback */
+    g_callback = callback;
+
+    /* Set g_tids_len */
+    g_tids_len = tids_len;
+
+    /* Allocate the output buffer */
+    g_output_array = zmalloc(sizeof(void*) * tids_len);
+
+    /* Initialize a semaphore that we will be waiting on for the threads
+    use pshared = 0 to indicate the semaphore is shared between the process's threads (and not between processes),
+    and value = 0 as the initial semaphore value. */
+    sem_init(&wait_for_threads_sem, 0, 0);
```

Comment:
The documentation indicates that while `sem_post` is async thread safe, `sem_init` and `sem_destroy` aren't.